### ISSUE TITLE
Support for paginated queries in JSON API

### ISSUE THREAD

**Issue by GeorgSchneider** [digital-asset/daml#6595]
Client feedback / feature request: For large ACS sizes it would be beneficial to provide a way to paginate query results so that this doesn't have to be done on the client-side. At some point loading the full contract set becomes a performance issue.

**Comment by AlexMartinezVila** [comment]
Will this pagination feature provide a sorting capability as well? If not, is there any plan for this?

**Comment by S11001001** [comment]
@meiersi-da suggests that an event_sequential_id, such as we used in #6658/#6372/#6215 for transaction streams, would satisfy this; the caveat being that our current schema denies any time-consistent set of pages, as one would expect ledger "now" to change as one is reading different pages, just as a query made 5 seconds later will return different results from the query 5 seconds before.

The client could use the last sequential ID as an opaque offset; this permits cheap next/previous page, but not cheap arbitrary page-jumping. I would suggest #6596 as a prerequisite.

**Comment by S11001001** [comment]
It's worth refining this, because different ideas of "pagination" mean different things. You may think you and someone else asking for "pagination" are asking for the same feature, when you really have very different requirements in mind.

One variant is the _complete partition_. It means that for every whole synchronous query made at time _t_ whose results are sufficiently large, there exists a set of requests, the first of which must be made at time _t_, such that appending the results of the requests yields the same contract set as in the original query. This is a pure performance optimization that can always be substituted for a single query with the same logical results.

Another variant is _next page now_. In this variant, when another page is requested, that request is considered to be a later "now" than when the previous page is requested. In other words, the invariant described for _complete partition_ cannot hold. This is more like what you would expect from a CRUD-style pagination: if you look at page 1 and someone deletes a record on page 2, you won't see that record on page 2 if you load after the deletion. Now "follows" you.

The _complete partition_ doesn't really fit into the semantic model that we present for the JSON API, nor does it fit into how we think about storing contracts. _Next page now_ does, because its core invariant is "every returned contract was an active contract at request time". So we are more likely to support some variant of _next page now_. The immediate prior comment describes one (SQL query store required) scheme for doing this.

That example demonstrates that other invariants clients can rely on _do_ arise from _next page now_ semantics, though the details of the invariants vary. In that example's case, you can expect that "no contract (by ID) occurs on more than one page".

Attempting to interpret _next page now_ as _complete partition_ always yields logical inconsistencies; again in this example, you may see duplicate occurrences of a contract _key_, as the initially-yielded contract might be archived and replaced before the later page load.

**Comment by stefanobaghino-da** [comment]
Needs scoping and design, see previous comment.

**Comment by RethinkLedgers** [comment]
We are eager to see a solution for JSON-API Pagination for the client solution we are building. Happy to contribute to the requirements gathering on this topic

**Comment by HarshMultani** [comment]
Hi @stefanobaghino-da,

Any update on this issue. I also need support for pagination and sorting in my usecase.

**Comment by S11001001** [comment]
One data point: user lists in the user management API #12610 use _next page now_, though I am unsure of the specific concurrent consistency guarantees they provide.

This is just a demonstration of the difference; the downstream use cases for contract queries and paging user lists are...different, to say the least, so it does not mean that the approach of this feature should (or should not) be similar in any way.

**Comment by meiersi-da** [comment]
@S11001001 : good point wrt the differences in requirements. For the user listing endpoint, each page is computed from a consistent snapshot of the list of all users; and consecutive pages are guaranteed to not overlap. It is though not possible to iterate through the pages to fetch a complete list of all users as-of a specific time, if the set of users is changing while the pages are being fetched.

A note in case you are going forward on designing pagination for the JSON-API. The infrastructure underlying the ACS endpoint is sufficiently powerful to implement pagination with snapshot consistency across different page fetches, as long as the snapshot is not pruned by the operator using ledger pruning. The necessary changes to the endpoints to expose pagination on the Ledger API have though not yet been implemented, nor scheduled.

**Comment by zeason** [comment]
Hi @meiersi-da, is there any progress in this feature request?

**Comment by meiersi-da** [comment]
> Hi @meiersi-da, is there any progress in this feature request?

As far as I know, not yet. @piyushbedi-da is the PM for this. Perhaps he knows more?

**Comment by mziolekda** [comment]
Closing, obsolete issue, old JSON API Server is no longer available in 3.x.
JSON Ledger API is now implemented as a transalation of the underlying gRPC. Support for pagination would have to be added there.
